diff --git a/.gitignore b/.gitignore
index c178dba..60af582 100644
--- a/.gitignore
+++ b/.gitignore
@@ -10,3 +10,8 @@ recordings
 *.swp
 *~
 noVNC-*.tgz
+
+#build
+[Oo]bj/
+[Bb]in/
+*.user
\ No newline at end of file
diff --git a/Properties/AssemblyInfo.cs b/Properties/AssemblyInfo.cs
new file mode 100644
index 0000000..79e92c2
--- /dev/null
+++ b/Properties/AssemblyInfo.cs
@@ -0,0 +1,52 @@
+n++// Copyright (c) 2016, FuseCP
+// FuseCP is distributed under the Creative Commons Share-alike license
+// 
+// FuseCP is a fork of WebsitePanel:
+// Copyright (c) 2014, Outercurve Foundation.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification,
+// are permitted provided that the following conditions are met:
+//
+// - Redistributions of source code must  retain  the  above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// - Redistributions in binary form  must  reproduce the  above  copyright  notice,
+//   this list of conditions  and  the  following  disclaimer in  the documentation
+//   and/or other materials provided with the distribution.
+//
+// - Neither  the  name  of  the  Outercurve Foundation  nor   the   names  of  its
+//   contributors may be used to endorse or  promote  products  derived  from  this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT  NOT  LIMITED TO, THE IMPLIED
+// WARRANTIES  OF  MERCHANTABILITY   AND  FITNESS  FOR  A  PARTICULAR  PURPOSE  ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+// ANY DIRECT, INDIRECT, INCIDENTAL,  SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO,  PROCUREMENT  OF  SUBSTITUTE  GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER  CAUSED AND ON
+// ANY  THEORY  OF  LIABILITY,  WHETHER  IN  CONTRACT,  STRICT  LIABILITY,  OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE)  ARISING  IN  ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("FuseCP.Providers.Virtualization.NoVNC")]
+[assembly: AssemblyProduct("FuseCP.Providers.Virtualization.NoVNC")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+// Setting ComVisible to false makes the types in this assembly not visible 
+// to COM components.  If you need to access a type in this assembly from 
+// COM, set the ComVisible attribute to true on that type.
+[assembly: ComVisible(false)]
+
+// The following GUID is for the ID of the typelib if this project is exposed to COM
+[assembly: Guid("959c2614-2d69-4e4f-9e77-bd868e5afd4b")]
diff --git a/FuseCP.Providers.Virtualization.NoVNC.csproj b/FuseCP.Providers.Virtualization.NoVNC.csproj
new file mode 100644
index 0000000..508a074
--- /dev/null
+++ b/FuseCP.Providers.Virtualization.NoVNC.csproj
@@ -0,0 +1,183 @@
+n++<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="12.0">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProductVersion>9.0.30729</ProductVersion>
+    <SchemaVersion>2.0</SchemaVersion>
+    <ProjectGuid>{D99EFB18-FFE7-45BB-8CA8-29336F3E8C68}</ProjectGuid>
+    <ProjectTypeGuids>{349c5851-65df-11da-9384-00065b846f21};{fae04ec0-301f-11d3-bf4b-00c04f79efbc}</ProjectTypeGuids>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>FuseCP.Providers.Virtualization.NoVNC</RootNamespace>
+    <AssemblyName>FuseCP.Providers.Virtualization.NoVNC</AssemblyName>
+    <FileUpgradeFlags>
+    </FileUpgradeFlags>
+    <OldToolsVersion>4.0</OldToolsVersion>
+    <UpgradeBackupLocation>
+    </UpgradeBackupLocation>
+    <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
+    <TargetFrameworkProfile />
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
+    <RestorePackages>true</RestorePackages>
+    <UseIISExpress>true</UseIISExpress>
+    <Use64BitIISExpress />
+    <IISExpressSSLPort />
+    <IISExpressAnonymousAuthentication />
+    <IISExpressWindowsAuthentication />
+    <IISExpressUseClassicPipelineMode />
+    <UseGlobalApplicationHostFile />
+  </PropertyGroup>
+  <PropertyGroup>
+    <PortalDirectory>..\..\Sources\FuseCP.WebPortal</PortalDirectory>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>portable</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>bin\$(Configuration)</OutputPath>
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
+    <WarningsAsErrors>618</WarningsAsErrors>
+    <PackageAsSingleFile>false</PackageAsSingleFile>
+    <FilesToIncludeForPublish>AllFilesInTheProject</FilesToIncludeForPublish>
+    <Prefer32Bit>false</Prefer32Bit>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>portable</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\$(Configuration)</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
+    <WarningsAsErrors>618</WarningsAsErrors>
+    <Prefer32Bit>false</Prefer32Bit>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="System" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="System.Web" />
+    <Reference Include="System.Web.Extensions" />
+    <Reference Include="System.Xml" />
+    <Reference Include="System.Configuration" />
+    <Reference Include="System.EnterpriseServices" />
+    <Reference Include="System.Xml.Linq" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="$(PortalDirectory)\..\VersionInfo.cs">
+      <Link>Properties\VersionInfo.cs</Link>
+    </Compile>
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="vnc.aspx.cs">
+      <DependentUpon>vnc.aspx</DependentUpon>
+      <SubType>ASPXCodeBehind</SubType>
+    </Compile>
+    <Compile Include="vnc.aspx.designer.cs">
+      <DependentUpon>vnc.aspx</DependentUpon>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <AppContent Include="app\**\*.*">
+      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
+    </AppContent>
+    <CoreContent Include="core\**\*.*">
+      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
+    </CoreContent>
+    <VendorContent Include="vendor\**\*.*">
+      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
+    </VendorContent>
+    <Content Include="app\fusecp.js" />
+    <Content Include="vnc.html">
+      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="vnc.aspx" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\Sources\FuseCP.EnterpriseServer.Base\FuseCP.EnterpriseServer.Base.csproj">
+      <Project>{c09ce910-f16b-48a1-b2cc-c99b8c1cf775}</Project>
+      <Name>FuseCP.EnterpriseServer.Base</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Sources\FuseCP.EnterpriseServer\WebServices\FuseCP.Build\FuseCP.EnterpriseServer.Client.csproj">
+      <Project>{df3d477e-741c-4871-b838-c1635e72474b}</Project>
+      <Name>FuseCP.EnterpriseServer.Client</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Sources\FuseCP.Providers.Base\FuseCP.Providers.Base.csproj">
+      <Project>{684c932a-6c75-46ac-a327-f3689d89eb42}</Project>
+      <Name>FuseCP.Providers.Base</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Sources\FuseCP.Web.Clients\FuseCP.Web.Clients.csproj">
+      <Project>{652639d5-59ce-4f42-8a5f-87cb1e812961}</Project>
+      <Name>FuseCP.Web.Clients</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Sources\FuseCP.WebPortal\DesktopModules\FuseCP\FuseCP.Portal.Modules.csproj">
+      <Project>{12232731-5c45-4ed6-98f8-d47abe728280}</Project>
+      <Name>FuseCP.Portal.Modules</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\..\Sources\FuseCP.WebPortal\FuseCP.WebPortal.csproj">
+      <Project>{c99efb18-ffe7-45bb-8ca8-29336f3e8c68}</Project>
+      <Name>FuseCP.WebPortal</Name>
+    </ProjectReference>
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Common\" />
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="package.json" />
+  </ItemGroup>
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
+  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
+       Other similar extension points exist, see Microsoft.Common.targets.
+  <Target Name="BeforeBuild">
+  </Target>-->
+  <Target Name="AfterBuild" DependsOnTargets="CopyContentToPortal">
+  </Target>
+  <Target Name="CopyContentToPortal">
+    <ItemGroup>
+      <OutputFiles Include="bin\$(Configuration)\FuseCP.Providers.Virtualization.NoVNC.*" />
+    </ItemGroup>
+    <MakeDir Directories="$(PortalDirectory)\novnc" Condition="!Exists('$(PortalDirectory)\novnc')" />
+    <Copy SourceFiles="@(Content)" DestinationFiles="@(Content->'$(PortalDirectory)\novnc\%(RecursiveDir)%(Filename)%(Extension)')" SkipUnchangedFiles="true" />
+    <Copy SourceFiles="@(AppContent)" DestinationFiles="@(AppContent->'$(PortalDirectory)\novnc\app\%(RecursiveDir)%(Filename)%(Extension)')" SkipUnchangedFiles="true" />
+    <Copy SourceFiles="@(CoreContent)" DestinationFiles="@(CoreContent->'$(PortalDirectory)\novnc\core\%(RecursiveDir)%(Filename)%(Extension)')" SkipUnchangedFiles="true" />
+    <Copy SourceFiles="@(VendorContent)" DestinationFiles="@(VendorContent->'$(PortalDirectory)\novnc\vendor\%(RecursiveDir)%(Filename)%(Extension)')" SkipUnchangedFiles="true" />
+    <Copy SourceFiles="@(OutputFiles)" DestinationFolder="$(PortalDirectory)\bin" />
+  </Target>
+  <Import Project="$(VSToolsPath)\WebApplications\Microsoft.WebApplication.targets" Condition="'$(VSToolsPath)' != ''" />
+  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v10.0\WebApplications\Microsoft.WebApplication.targets" Condition="false" />
+  <PropertyGroup>
+    <IncludeSetAclProviderOnDestination>False</IncludeSetAclProviderOnDestination>
+    <ExcludeFilesFromPackage>True</ExcludeFilesFromPackage>
+  </PropertyGroup>
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
+  <ProjectExtensions>
+    <VisualStudio>
+      <FlavorProperties GUID="{349c5851-65df-11da-9384-00065b846f21}">
+        <WebProjectProperties>
+          <UseIIS>True</UseIIS>
+          <AutoAssignPort>True</AutoAssignPort>
+          <DevelopmentServerPort>0</DevelopmentServerPort>
+          <DevelopmentServerVPath>/</DevelopmentServerVPath>
+          <IISUrl>http://localhost:60343/</IISUrl>
+          <NTLMAuthentication>False</NTLMAuthentication>
+          <UseCustomServer>False</UseCustomServer>
+          <CustomServerUrl>
+          </CustomServerUrl>
+          <SaveServerSettingsInUserFile>False</SaveServerSettingsInUserFile>
+        </WebProjectProperties>
+      </FlavorProperties>
+    </VisualStudio>
+  </ProjectExtensions>
+</Project>
\ No newline at end of file
diff --git a/app/fusecp.js b/app/fusecp.js
new file mode 100644
index 0000000..9c977d9
--- /dev/null
+++ b/app/fusecp.js
@@ -0,0 +1,439 @@
+n++/*
+* FuseCP Utility functions for noVNC
+* Copyright (c) 2024, FuseCP
+* FuseCP is distributed under the Creative Commons Share-alike license
+*/
+
+import * as WebUtil from "./webutil.js";
+
+export default function FCPUI(UI) {
+	this.consoletype = WebUtil.getQueryVar('console');
+	this.vmid = WebUtil.getQueryVar('vmid');
+	this.vmname = WebUtil.getQueryVar('vmname');
+	this.nodename = WebUtil.getQueryVar('node');
+	this.resize = WebUtil.getQueryVar('resize');
+	this.lastFBWidth = undefined;
+	this.lastFBHeight = undefined;
+	this.sizeUpdateTimer = undefined;
+	this.UI = UI;
+
+	var baseUrl = '/nodes/' + this.nodename;
+	var url;
+	var params = { websocket: 1 };
+	var title;
+
+	switch (this.consoletype) {
+		case 'kvm':
+			baseUrl += '/qemu/' + this.vmid;
+			url = baseUrl + '/vncproxy';
+			title = "VM " + this.vmid;
+			if (this.vmname) {
+				title += " ('" + this.vmname + "')";
+			}
+			break;
+		case 'lxc':
+			baseUrl += '/lxc/' + this.vmid;
+			url = baseUrl + '/vncproxy';
+			title = "CT " + this.vmid;
+			if (this.vmname) {
+				title += " ('" + this.vmname + "')";
+			}
+			break;
+		case 'shell':
+			url = baseUrl + '/vncshell';
+			title = "node '" + this.nodename + "'";
+			break;
+		case 'upgrade':
+			url = baseUrl + '/vncshell';
+			params.upgrade = 1;
+			title = 'System upgrade on node ' + this.nodename;
+			break;
+		default:
+			//throw 'implement me';
+			break;
+	}
+
+	if (this.resize == 'scale' &&
+		(this.consoletype === 'lxc' || this.consoletype === 'shell')) {
+		var size = this.getFBSize();
+		params.width = size.width;
+		params.height = size.height;
+	}
+
+	this.baseUrl = baseUrl;
+	this.url = url;
+	this.params = params;
+	document.title = title;
+};
+
+FCPUI.prototype = {
+	urlEncode: function (object) {
+		var i, value, params = [];
+
+		for (i in object) {
+			if (object.hasOwnProperty(i)) {
+				value = object[i];
+				if (value === undefined) value = '';
+				params.push(encodeURIComponent(i) + '=' + encodeURIComponent(String(value)));
+			}
+		}
+
+		return params.join('&');
+	},
+
+	API2Request: function (reqOpts) {
+		var me = this;
+
+		reqOpts.method = reqOpts.method || 'GET';
+
+		var xhr = new XMLHttpRequest();
+
+		xhr.onload = function () {
+			var scope = reqOpts.scope || this;
+			var result;
+			var errmsg;
+
+			if (xhr.readyState === 4) {
+				var ctype = xhr.getResponseHeader('Content-Type');
+				if (xhr.status === 200) {
+					if (ctype.match(/application\/json;/)) {
+						result = JSON.parse(xhr.responseText);
+					} else {
+						errmsg = 'got unexpected content type ' + ctype;
+					}
+				} else {
+					errmsg = 'Error ' + xhr.status + ': ' + xhr.statusText;
+				}
+			} else {
+				errmsg = 'Connection error - server offline?';
+			}
+
+			if (errmsg !== undefined) {
+				if (reqOpts.failure) {
+					reqOpts.failure.call(scope, errmsg);
+				}
+			} else {
+				if (reqOpts.success) {
+					reqOpts.success.call(scope, result);
+				}
+			}
+			if (reqOpts.callback) {
+				reqOpts.callback.call(scope, errmsg === undefined);
+			}
+		}
+
+		var data = me.urlEncode(reqOpts.params || {});
+
+		if (reqOpts.method === 'GET') {
+			xhr.open(reqOpts.method, "/api2/json" + reqOpts.url + '?' + data);
+		} else {
+			xhr.open(reqOpts.method, "/api2/json" + reqOpts.url);
+		}
+		xhr.setRequestHeader('Cache-Control', 'no-cache');
+		if (reqOpts.method === 'POST' || reqOpts.method === 'PUT') {
+			xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
+			xhr.setRequestHeader('CSRFPreventionToken', PVE.CSRFPreventionToken);
+			xhr.send(data);
+		} else if (reqOpts.method === 'GET') {
+			xhr.send();
+		} else {
+			throw "unknown method";
+		}
+	},
+
+	pve_detect_migrated_vm: function () {
+		var me = this;
+		if (me.consoletype === 'kvm') {
+			// try to detect migrated VM
+			me.API2Request({
+				url: '/cluster/resources',
+				method: 'GET',
+				success: function (result) {
+					var list = result.data;
+					list.every(function (item) {
+						if (item.type === 'qemu' && item.vmid == me.vmid) {
+							var url = "?" + me.urlEncode({
+								console: me.consoletype,
+								novnc: 1,
+								vmid: me.vmid,
+								vmname: me.vmname,
+								node: item.node,
+								resize: me.resize
+							});
+							location.href = url;
+							return false; // break
+						}
+						return true;
+					});
+				}
+			});
+		} else if (me.consoletype === 'lxc') {
+			// lxc restart migration can take a while,
+			// so we need to find out if we are really migrating
+			var migrating;
+			var check = setInterval(function () {
+				if (migrating === undefined ||
+					migrating === true) {
+					// check (again) if migrating
+					me.UI.showStatus('Waiting for connection...', 'warning', 5000);
+					me.API2Request({
+						url: me.baseUrl + '/config',
+						method: 'GET',
+						success: function (result) {
+							var lock = result.data.lock;
+							if (lock == 'migrate') {
+								migrating = true;
+								me.UI.showStatus('Migration detected, waiting...', 'warning', 5000);
+							} else {
+								migrating = false;
+							}
+						},
+						failure: function () {
+							migrating = false;
+						}
+					});
+				} else {
+					// not migrating any more
+					me.UI.showStatus('Connection resumed', 'warning');
+					clearInterval(check);
+					me.API2Request({
+						url: '/cluster/resources',
+						method: 'GET',
+						success: function (result) {
+							var list = result.data;
+							list.every(function (item) {
+								if (item.type === 'lxc' && item.vmid == me.vmid) {
+									var url = "?" + me.urlEncode({
+										console: me.consoletype,
+										novnc: 1,
+										vmid: me.vmid,
+										vmname: me.vmname,
+										node: item.node,
+										resize: me.resize
+									});
+									location.href = url;
+									return false; // break
+								}
+								return true;
+							});
+						}
+					});
+				}
+			}, 5000);
+		}
+
+	},
+
+	pve_vm_command: function (cmd, params, reload) {
+		var me = this;
+		var baseUrl;
+		var confirmMsg = "";
+
+		switch (cmd) {
+			case "start":
+				reload = 1;
+			case "shutdown":
+			case "stop":
+			case "reset":
+			case "suspend":
+			case "resume":
+				confirmMsg = "Do you really want to " + cmd + " VM/CT {0}?";
+				break;
+			case "reload":
+				location.reload();
+				break;
+			default:
+				throw "implement me " + cmd;
+		}
+
+		confirmMsg = confirmMsg.replace('{0}', me.vmid);
+
+		if (confirmMsg !== "" && confirm(confirmMsg) !== true) {
+			return;
+		}
+
+		me.UI.closePVECommandPanel();
+
+		if (me.consoletype === 'kvm') {
+			baseUrl = '/nodes/' + me.nodename + '/qemu/' + me.vmid;
+		} else if (me.consoletype === 'lxc') {
+			baseUrl = '/nodes/' + me.nodename + '/lxc/' + me.vmid;
+		} else {
+			throw "unknown VM type";
+		}
+
+		me.API2Request({
+			params: params,
+			url: baseUrl + "/status/" + cmd,
+			method: 'POST',
+			failure: function (msg) {
+				if (cmd === 'start' && msg.match(/already running/) !== null) {
+					// we wanted to start, but it was already running, so
+					// reload anyway
+					me.UI.showStatus("VM command '" + cmd + "' successful", 'normal');
+					setTimeout(function () {
+						location.reload();
+					}, 1000);
+				} else {
+					me.UI.showStatus(msg, 'warning');
+				}
+			},
+			success: function () {
+				me.UI.showStatus("VM command '" + cmd + "' successful", 'normal');
+				if (reload) {
+					setTimeout(function () {
+						location.reload();
+					}, 1000);
+				};
+			}
+		});
+	},
+
+	addPVEHandlers: function () {
+		var me = this;
+		document.getElementById('pve_commands_button')
+			.addEventListener('click', me.UI.togglePVECommandPanel);
+
+		// show/hide the buttons
+		document.getElementById('noVNC_disconnect_button')
+			.classList.add('noVNC_hidden');
+		if (me.consoletype === 'kvm') {
+			document.getElementById('noVNC_clipboard_button')
+				.classList.add('noVNC_hidden');
+		}
+
+		if (me.consoletype === 'shell' || me.consoletype === 'upgrade') {
+			document.getElementById('pve_commands_button')
+				.classList.add('noVNC_hidden');
+		}
+
+		// add command logic
+		var commandArray = [
+			{ cmd: 'start', kvm: 1, lxc: 1 },
+			{ cmd: 'stop', kvm: 1, lxc: 1 },
+			{ cmd: 'shutdown', kvm: 1, lxc: 1 },
+			{ cmd: 'suspend', kvm: 1 },
+			{ cmd: 'resume', kvm: 1 },
+			{ cmd: 'reset', kvm: 1 },
+			{ cmd: 'reload', kvm: 1, lxc: 1, shell: 1 },
+		];
+
+		commandArray.forEach(function (item) {
+			var el = document.getElementById('pve_command_' + item.cmd);
+			if (!el) {
+				return;
+			}
+
+			if (item[me.consoletype] === 1) {
+				el.onclick = function () {
+					me.pve_vm_command(item.cmd);
+				};
+			} else {
+				el.classList.add('noVNC_hidden');
+			}
+		});
+	},
+
+	getFBSize: function () {
+		var oh;
+		var ow;
+
+		if (window.innerHeight) {
+			oh = window.innerHeight;
+			ow = window.innerWidth;
+		} else if (document.documentElement &&
+			document.documentElement.clientHeight) {
+			oh = document.documentElement.clientHeight;
+			ow = document.documentElement.clientWidth;
+		} else if (document.body) {
+			oh = document.body.clientHeight;
+			ow = document.body.clientWidth;
+		} else {
+			throw "can't get window size";
+		}
+
+		return { width: ow, height: oh };
+	},
+
+	pveStart: function (callback) {
+		var me = this;
+		me.API2Request({
+			url: me.url,
+			method: 'POST',
+			params: me.params,
+			success: function (result) {
+				var wsparams = me.urlEncode({
+					port: result.data.port,
+					vncticket: result.data.ticket
+				});
+
+				document.getElementById('noVNC_password_input').value = result.data.ticket;
+				me.UI.forceSetting('path', 'api2/json' + me.baseUrl + '/vncwebsocket' + "?" + wsparams);
+
+				callback();
+			},
+			failure: function (msg) {
+				me.UI.showStatus(msg, 'error');
+			}
+		});
+	},
+
+	updateFBSize: function (rfb, width, height) {
+		var me = this;
+		try {
+			// Note: window size must be even number for firefox
+			me.lastFBWidth = Math.floor((width + 1) / 2) * 2;
+			me.lastFBHeight = Math.floor((height + 1) / 2) * 2;
+
+			if (me.sizeUpdateTimer !== undefined) {
+				clearInterval(me.sizeUpdateTimer);
+			}
+
+			var update_size = function () {
+				var clip = me.UI.getSetting('view_clip');
+				var resize = me.UI.getSetting('resize');
+				var autoresize = me.UI.getSetting('autoresize');
+				if (clip || resize === 'scale' || !autoresize) {
+					return;
+				}
+
+				// we do not want to resize if we are in fullscreen
+				if (document.fullscreenElement || // alternative standard method
+					document.mozFullScreenElement || // currently working methods
+					document.webkitFullscreenElement ||
+					document.msFullscreenElement) {
+					return;
+				}
+
+				var oldsize = me.getFBSize();
+				var offsetw = me.lastFBWidth - oldsize.width;
+				var offseth = me.lastFBHeight - oldsize.height;
+				if (offsetw !== 0 || offseth !== 0) {
+					//console.log("try resize by " + offsetw + " " + offseth);
+					try {
+						window.resizeBy(offsetw, offseth);
+					} catch (e) {
+						console.log('resizing did not work', e);
+					}
+				}
+			};
+
+			update_size();
+			me.sizeUpdateTimer = setInterval(update_size, 1000);
+
+		} catch (e) {
+			console.log(e);
+		}
+	},
+
+	updateSessionSize: function (e) {
+		var rfb = e.detail.rfb;
+		var width = e.detail.width;
+		var height = e.detail.height;
+		UI.FCP.updateFBSize(rfb, width, height);
+
+		UI.applyResizeMode();
+		UI.updateViewClip();
+	},
+
+};
\ No newline at end of file
diff --git a/noVNC.sln b/noVNC.sln
new file mode 100644
index 0000000..d080b91
--- /dev/null
+++ b/noVNC.sln
@@ -0,0 +1,25 @@
+n++
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 17
+VisualStudioVersion = 17.5.002.0
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "FuseCP.Providers.Virtualization.NoVNC", "FuseCP.Providers.Virtualization.NoVNC.csproj", "{1FEFD6D3-B584-4AF5-85EA-AA74DF9714A7}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
+		Release|Any CPU = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{1FEFD6D3-B584-4AF5-85EA-AA74DF9714A7}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{1FEFD6D3-B584-4AF5-85EA-AA74DF9714A7}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{1FEFD6D3-B584-4AF5-85EA-AA74DF9714A7}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{1FEFD6D3-B584-4AF5-85EA-AA74DF9714A7}.Release|Any CPU.Build.0 = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {37DF1737-B5CB-4C67-A83E-564D5C5EDDFC}
+	EndGlobalSection
+EndGlobal
diff --git a/vnc.aspx b/vnc.aspx
new file mode 100644
index 0000000..036dfa9
--- /dev/null
+++ b/vnc.aspx
@@ -0,0 +1,47 @@
+n++<%@ Page Language="C#" AutoEventWireup="true" CodeBehind="vnc.aspx.cs" Inherits="FuseCP.Providers.Virtualization.NoVNC.vnc"
+	EnableTheming="False" StylesheetTheme="" Theme="" %>
+
+<DOCTYPE html>
+<html lang="en" class="noVNC_loading">
+<head runat="server">
+
+	<asp:Literal ID="headHtml" runat="server" />
+
+	<script type="module" crossorigin="anonymous" src="app/error-handler.js"></script>
+	<%--<script type="module" crossorigin="anonymous" src="app/ui.js"></script>--%>
+	<%--<script type="module" crossorigin="anonymous" src="app/fusecp.js"></script>--%>
+</head>
+<body style="border: 0; margin: 0; padding: 0;">
+
+	<asp:Literal ID="bodyHtml" runat="server" />
+
+	<form id="form1" runat="server" autocomplete="off"></form>
+
+	<script type="module">
+
+		import UI from "./app/ui.js";
+		import FCPUI from "./app/fusecp.js";
+
+		UI.setupFuseCP = function () {
+
+			UI.updateViewClip();
+
+			UI.forceSetting('path', <%= Path %>);
+			UI.forceSetting('port', <%= Port %>);
+			UI.forceSetting('host', <%= Host %>);
+
+			//UI.forceSetting('encrypt', true);
+			UI.forceSetting('autoresize', true);
+			UI.addSettingChangeHandler('autoresize');
+
+			UI.connect(null, <%= Password %>);
+		};
+
+		UI.FCP = new FCPUI(UI);
+
+		UI.setup();
+
+	</script>
+
+</body>
+</html>
diff --git a/vnc.aspx.cs b/vnc.aspx.cs
new file mode 100644
index 0000000..0369353
--- /dev/null
+++ b/vnc.aspx.cs
@@ -0,0 +1,81 @@
+n++using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+using System.Web.UI;
+using System.Web.UI.WebControls;
+using System.IO;
+using FuseCP.Providers;
+using FuseCP.Providers.OS;
+using FuseCP.Portal;
+using FuseCP.WebPortal;
+using System.Text.RegularExpressions;
+
+
+namespace FuseCP.Providers.Virtualization.NoVNC
+{
+    public partial class vnc : System.Web.UI.Page
+    {
+        public const string DEFAULT_PAGE = "~/Default.aspx";
+        public const string PAGE_ID_PARAM = "pid";
+        public const string ITEM_ID_PARAM = "item";
+
+        public string Host { get; set; }
+        public int Port { get; set; }
+        public string Path { get; set; }
+        public string VmName { get; set; }
+        public string Password { get; set; }
+
+        public string EscapedJSString(string text)
+        {
+            text = text.Replace(@"\", @"\\")
+                .Replace(@"'", @"\'");
+            return $"'{text}'";
+        }
+        protected void Page_Load(object sender, EventArgs e)
+        {
+            var loggedIn = PanelSecurity.LoggedUser != null;
+            if (!loggedIn)
+            {
+                // redirect to login page
+                string returnUrl = Request.RawUrl;
+                Response.Redirect(DEFAULT_PAGE + "?" + PAGE_ID_PARAM + "=" +
+                    PortalConfiguration.SiteSettings["LoginPage"] + "&ReturnUrl=" + Server.UrlEncode(returnUrl));
+            }
+
+            var baseUrl = Regex.Match(Request.Path, "(?<=^/).*(?=/.*?.aspx)")?.Value;
+
+            Host = EscapedJSString(Request.Url.Host);
+            Port = Request.Url.Port;
+
+            var item = Request.QueryString[ITEM_ID_PARAM];
+            if (item != null)
+            {
+                var itemId = int.Parse(item);
+                var vm = Portal.ES.Services.Packages.GetPackageItem(itemId) as VirtualMachine;
+                if (vm != null) Page.Title = VmName = vm.Name;
+                else Page.Title = VmName = "noVNC";
+
+                var vncCredentials = Portal.ES.Services.Proxmox.GetPveVncCredentials(itemId);
+                if (vncCredentials != null)
+                {
+                    Path = EscapedJSString($"{baseUrl}/websocket?user={Uri.EscapeDataString(PanelSecurity.LoggedUser.Username)}&item={itemId}&vncpassword={Uri.EscapeDataString(vncCredentials.Password)}");
+                    Password = EscapedJSString(vncCredentials.Password);
+
+                    //var url = $"{baseUrl}/vnc.html?autoconnect=true&host={Request.Url.Host}&port={Request.Url.Port}&password={Uri.EscapeDataString(vncCredentials.Password)}&path={Uri.EscapeDataString(path)}";
+                    //Response.Redirect(url);
+                }
+            }
+
+            var htmlText = File.ReadAllText(Server.MapPath("~/novnc/vnc.html"));
+
+            var match = Regex.Match(htmlText, @"<head[^/>]*?>(?<head>.*?)</head>.*?<body[^/>]*?>(?<body>.*?)</body>", RegexOptions.Singleline | RegexOptions.IgnoreCase);
+            string head = match.Groups["head"].Value;
+            string body = match.Groups["body"].Value;
+            bodyHtml.Text = body;
+            // remove title & scripts from head
+            head = Regex.Replace(head, @"(?:<title>.*?</title>|<script[^>]*(?<!/)>.*?</script>|<script[^>]*?/>)(?:[ \t]*?\r?\n)?", "", RegexOptions.Singleline | RegexOptions.IgnoreCase);
+            headHtml.Text = head;
+        }
+    }
+}
diff --git a/vnc.aspx.designer.cs b/vnc.aspx.designer.cs
new file mode 100644
index 0000000..4cb1217
--- /dev/null
+++ b/vnc.aspx.designer.cs
@@ -0,0 +1,44 @@
+n++//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by a tool.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated. 
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace FuseCP.Providers.Virtualization.NoVNC
+{
+
+
+    public partial class vnc
+    {
+
+        /// <summary>
+        /// headHtml control.
+        /// </summary>
+        /// <remarks>
+        /// Auto-generated field.
+        /// To modify move field declaration from designer file to code-behind file.
+        /// </remarks>
+        protected global::System.Web.UI.WebControls.Literal headHtml;
+
+        /// <summary>
+        /// bodyHtml control.
+        /// </summary>
+        /// <remarks>
+        /// Auto-generated field.
+        /// To modify move field declaration from designer file to code-behind file.
+        /// </remarks>
+        protected global::System.Web.UI.WebControls.Literal bodyHtml;
+
+        /// <summary>
+        /// form1 control.
+        /// </summary>
+        /// <remarks>
+        /// Auto-generated field.
+        /// To modify move field declaration from designer file to code-behind file.
+        /// </remarks>
+        protected global::System.Web.UI.HtmlControls.HtmlForm form1;
+    }
+}
